Update on-ramp providers and add purchase method screen

simoarpe opened the pull request:
## Resolves https://github.com/brave/brave-browser/issues/23316 (Update on-ramp methods)
## Resolves https://github.com/brave/brave-browser/issues/27996 (Switch to new API `getBuyUrlV1`)

Implements a new screen to select a purchase method containing Ramp.Network, Sardine and Transak.
Adopted [new API `getBuyUrlV1`](https://github.com/brave/brave-core/blob/03673e48c89965d88856204a78d3e1365645dfe0/components/brave_wallet/browser/asset_ratio_service.cc#L180).

* ⚠️ Sardine will appear only in US locales.
* Ramp.Network needs modified symbols obtained by `mapToRampNetworkSymbol`.
* New core method that accepts an array of providers, and returns the full list.
* Implemented `removeDuplicates` to avoid showing duplication in the `EditVisibleAssetsBottomSheetDialogFragment` UI.
* Update Sardine with new logo.


<img width="515" alt="Screenshot 2023-01-31 at 1 29 34 PM" src="https://user-images.githubusercontent.com/3308503/215760233-4371b093-9324-416e-b14d-2a073a695f75.png">

<img width="515" alt="Screenshot 2023-01-31 at 1 31 30 PM" src="https://user-images.githubusercontent.com/3308503/215760537-13deae14-e668-4987-be5c-3eebeccd04cc.png">

## Demo

https://user-images.githubusercontent.com/3308503/215813132-85e15ad3-97ce-48e2-8e5b-1aa33fb52eef.mov

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Open the wallet section
- Tap on FAB icon and select "Buy"
- Tap on "Select purchase method" button
- Observe the new screen "Select purchase method"
- Verify that tapping on the "Buy with Ramp" button opens a new page on Ramp network
- Verify that tapping on the "Buy with Sardine" button opens a new page on Sardine
- Verify that tapping on the "Buy with Transak" button opens a new page on Transak

Pavneet-Sing commented:
I believe the primary task with on-ramp implementation is to expand the support by adding additional Ramp networks(Sardine etc). The UI is good to have but it was previously required due to multiple options: ramp and wyre. 
As of now, it's just an extra step to select the only one option available.
Isn't it better to keep the previous implementation to use ramp as a default option for buy?
Is there any plan to add more network options to this screen?
cc: @jamesmudgett @SergeyZhukovsky

simoarpe commented:
Is there any description of the actual changes required? I copied the iOS flow but couldn't find any other differences other than the extra screen "Select purchase method".

SergeyZhukovsky commented on android/brave_java_resources.gni:
```diff
@@ -175,6 +175,7 @@ brave_java_resources = [
   "java/res/drawable-nodpi/ic_quick_action_search_and_bookmark_widget_preview.png",
   "java/res/drawable-nodpi/loader0.png",
   "java/res/drawable-nodpi/loader0_orange.png",
+  "java/res/drawable-nodpi/ramp.png",
```
could we have it in svg format? We are going to convert all pictures to vector graphic soon. That will decrease resources size plus they are scalable. It doesn't make sense to add anything new via png.

simoarpe replied:
Extracted from iOS, so the only option available was PNG. I'm going to try to manually convert this.

SergeyZhukovsky replied:
Or ask design team for a help

simoarpe replied:
The results were low quality, I'll ask the design team for some help.

SergeyZhukovsky commented:
> I believe the primary task with on-ramp implementation is to expand the support by adding additional Ramp networks(Sardine etc). The UI is good to have but it was previously required due to multiple options: ramp and wyre. As of now, it's just an extra step to select the only one option available. Isn't it better to keep the previous implementation to use ramp as a default option for buy? Is there any plan to add more network options to this screen? cc: @jamesmudgett @SergeyZhukovsky

I would say there are plans to add more, but @jamesmudgett would know better.

simoarpe commented:
@Pavneet-Sing, here is a recording of the iOS flow that I copied. Compiled a debug version from the latest changes (Pulled the latest changes and recompiled right now just to be extra sure). The only difference seems to be this extra screen. Tried on both Solana and Ethereum networks. Let me know I'm missing something obvious.

https://user-images.githubusercontent.com/3308503/214850464-37d18142-0a77-4a51-8d0c-677075b3759e.mov

simoarpe commented:
Update: I've just had a call with @jamesmudgett discussing the changes required. I'm going to ping you again once this PR is ready to review.

Pavneet-Sing commented:
> @Pavneet-Sing, here is a recording of the iOS flow that I copied. Compiled a debug version from the latest changes (Pulled the latest changes and recompiled right now just to be extra sure). The only difference seems to be this extra screen. Tried on both Solana and Ethereum networks. Let me know I'm missing something obvious.
> 
>  Screen.Recording.2023-01-26.at.2.41.49.PM.mov

IOS sardine support: https://github.com/brave/brave-ios/blob/development/Sources/BraveWallet/Crypto/Stores/BuyTokenStore.swift#L24 but it is for US only: https://github.com/brave/brave-ios/blob/development/Sources/BraveWallet/Crypto/Stores/BuyTokenStore.swift#L160

Additionally, Desktop has `Transak` as well

![Screenshot from 2023-01-26 21-53-12](https://user-images.githubusercontent.com/11755381/214890868-cebf6bb5-877a-478a-8815-2a4923dce605.png)

SergeyZhukovsky commented on android/java/org/chromium/chrome/browser/crypto_wallet/activities/SelectPurchaseMethodActivity.java:
```diff
@@ -0,0 +1,119 @@
+        mTransakButton = findViewById(R.id.purchase_method_btn_transak);
+
+        if (mOnRampModel != null) {
+            mOnRampModel.getBuyUrl(OnRampProvider.RAMP, mChainId, mFrom, mRampNetworkSymbol,
```
we can't access mojo calls in that place. They do calls in native libraries that are initialised for each Activity in chromium. Check other Activities, every single access to native libs has to be done only when we know that native is initialised. Otherwise it could be a crash.
```
@Override
public void finishNativeInitialization() {
        super.finishNativeInitialization();
        // Your code starts here
}
```

simoarpe replied:
Oh, right! Changes on their way.

SergeyZhukovsky commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,255 @@
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
```
`paddingTop` and `paddingBottom` could be combined to `paddingVertical`

SergeyZhukovsky commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,255 @@
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
```
`paddingVertical`

SergeyZhukovsky commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,255 @@
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
```
`paddingVertical`

SergeyZhukovsky commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,255 @@
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
```
`paddingVertical`

SergeyZhukovsky commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,255 @@
+        android:paddingTop="8dp"
+        android:paddingBottom="8dp"
+        android:paddingStart="14dp"
+        android:paddingEnd="14dp"
```
`paddingVertical` and `paddingHorizontal`

simoarpe commented:
Review feedback addressed @SergeyZhukovsky.

SergeyZhukovsky reviewed (approved):
++

Pavneet-Sing commented:
I am noticing a small delay to show Sardine option that can be displayed instantly. Just create a method to check if sardine is enabled and show it instantly. This will also simplify `getBuyUrl`: https://github.com/brave/brave-core/blob/46bc7114e0c622ffb70232d43a51f723224297b8/android/java/org/chromium/chrome/browser/app/domain/OnRampModel.java#L27-L37

[sardine-delay.webm](https://user-images.githubusercontent.com/11755381/215972411-3669e343-df2c-4b84-b7ee-f4e2746db1e8.webm)

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+                android:layout_height="wrap_content"
+                android:orientation="horizontal">
+
+                <TextView
+                    android:id="@+id/select_purchase_method_title"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:textSize="20sp"
+                    android:textColor="@color/wallet_text_color"
+                    android:gravity="center_vertical"
+                    android:layout_marginEnd="20dp"
+                    android:text="@string/wallet_buy_mainnet_button_text"/>
```
Always use styles and theme , e.g. `BraveWalletTextView`

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:textSize="14sp"
+            android:layout_marginStart="16dp"
+            android:layout_marginEnd="16dp"
```
nit : `layout_marginHorizontal`

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+        <TextView
+            android:id="@+id/select_options_text"
+            style="@style/BraveWalletTextView"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:textSize="14sp"
```
not required

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+        android:id="@+id/ramp_network_container"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:paddingTop="8dp"
+        android:paddingBottom="8dp"
+        android:paddingStart="14dp"
+        android:paddingEnd="14dp"
+        android:layout_marginStart="14dp"
+        android:layout_marginEnd="14dp"
+        android:layout_marginBottom="8dp"
```
nit: use horizontal and vertical attributes

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+                android:paddingVertical="8dp"
+                android:text="@string/brave_wallet_ramp_network_title"
+                android:textColor="@color/wallet_text_color"
+                android:textSize="17sp" />
```
use even values

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+            <TextView
+                style="@style/BraveWalletTextViewTitle"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
```
nit: not required `android:layout_height`

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+            <TextView
+                style="@style/BraveWalletTextView"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
```
nit: not required `android:layout_height`

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+                android:gravity="start"
+                android:paddingVertical="8dp"
+                android:text="@string/brave_wallet_buy_crypto_visa_mastercard"
+                android:textColor="@color/wallet_text_color"
```
any reason why `android:textColor="wallet_text_color` is needed for all the textviews? seems like a CP error?

Pavneet-Sing replied:
Because `BraveWalletTextView` has basic colour (per figma), size and dimension support 

I have added some styles to reuse and feel free to add more if something can be reused. We should always prefer to use these styles. These style attributes can be overridden in XML if we need to do a single/small change etc.

https://github.com/brave/brave-core/blob/master/android/java/res/values/brave_styles.xml#L64-L70

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+        android:id="@+id/sardine_container"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:paddingTop="8dp"
+        android:paddingBottom="8dp"
+        android:paddingStart="14dp"
+        android:paddingEnd="14dp"
+        android:layout_marginStart="14dp"
+        android:layout_marginEnd="14dp"
+        android:layout_marginBottom="8dp"
```
nit: use vertical and horizontal attributes

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+            <TextView
+                style="@style/BraveWalletTextViewTitle"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingVertical="8dp"
+                android:text="@string/brave_wallet_sardine_title"
+                android:textColor="@color/wallet_text_color"
+                android:textSize="17sp" />
```
nit: same as above changes

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+                android:textSize="17sp" />
+
+            <TextView
+                style="@style/BraveWalletTextView"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingVertical="8dp"
+                android:text="@string/brave_wallet_instant_buy"
+                android:textColor="@color/wallet_text_color"
+                android:textSize="14sp" />
```
nit: same as above changes

Pavneet-Sing commented on android/java/res/layout/activity_select_purchase_method.xml:
```diff
@@ -0,0 +1,249 @@
+        android:baselineAligned="false"
+        android:visibility="gone">
+
+        <ImageView
+            android:layout_width="74dp"
+            android:layout_height="48dp"
+            android:layout_gravity="top"
+            android:paddingTop="8dp"
+            android:paddingEnd="14dp"
+            app:srcCompat="@drawable/transak_logo"
+            android:contentDescription="@null" />
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+
+            <TextView
+                style="@style/BraveWalletTextViewTitle"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
+                android:text="@string/brave_wallet_transak_title"
+                android:textColor="@color/wallet_text_color"
+                android:textSize="17sp" />
+
+            <TextView
+                style="@style/BraveWalletTextView"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="center"
+                android:gravity="start"
+                android:paddingTop="8dp"
+                android:paddingBottom="8dp"
+                android:text="@string/brave_wallet_instant_buy"
+                android:textColor="@color/wallet_text_color"
+                android:textSize="14sp" />
```
nit: same as above changes

Pavneet-Sing commented:
Looks good though I think there are two more issues that needs to be addressed

1. Hide unsupported providers from Purchase screen as not all tokens are supported by providers. e.g Tether USD is not supported by Sardine and selection of Sardine with Tether will show nothing or error

[buy-unsupported-providers.webm](https://user-images.githubusercontent.com/11755381/216082798-c8731bfa-d2d3-4fa4-ba2c-b1871e226928.webm)

For this, you will need to use [isBuySupported](https://github.com/brave/brave-core/commit/0a1b935a88e59bd134b7416abff2331572583403#diff-0206db739051bba6a151ad12e7ff41bc4fc3b6d8e72ef4719fd34f45bfeecdd8)

2. The buy option should be shown on AssetDetails screen and for that, need to update `isBuySupported` or something similar to check against the support tokens from all providers.

[missing-buy-support.webm](https://user-images.githubusercontent.com/11755381/216083327-7ccede60-115b-4f21-88fc-5fd2fd856046.webm)

simoarpe commented on android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java:
```diff
@@ -324,21 +335,10 @@ public void setAccountInfosFromKeyRingModel(
         mNetworkModel.setAccountInfosFromKeyRingModel(accountInfosFromKeyRingModel);
     }
 
-    // TODO: Move to BuyModel class
-    public void isBuySupported(NetworkInfo selectedNetwork, String assetSymbol,
-            String contractAddress, String chainId, Callback1<Boolean> callback) {
-        TokenUtils.getBuyTokensFiltered(
-                mBlockchainRegistry, selectedNetwork, TokenUtils.TokenType.ALL, tokens -> {
-                    callback.call(JavaUtils.includes(tokens,
-                            iToken
-                            -> AssetUtils.Filters.isSameToken(
-                                    iToken, assetSymbol, contractAddress, chainId)));
-                });
-    }
```
Moved this method into `BuyModel` class.

simoarpe commented on android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:
```diff
@@ -971,18 +971,11 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
                 }
             } else if (mActivityType == ActivityType.BUY) {
-                assert mBlockchainRegistry != null;
-                String symbol = AssetUtils.mapToRampNetworkSymbol(mCurrentBlockchainToken);
```
`mapToRampNetworkSymbol` is used only for `RAMP_NETWORK`. Moved the logic into `BuyModel` class.

simoarpe commented:
@Pavneet-Sing , feedback addressed. 
Asking also @yrliou for a review of the Core part.

@yrliou Context: needed a way to fetch buy tokens from multiple on-ramp providers. Solution: implemented `GetProvidersBuyTokens`.

Pavneet-Sing commented:
Search on buy is showing duplicate tokens.

[duplicate-search-items.webm](https://user-images.githubusercontent.com/11755381/217195462-eea62a0a-948e-4fce-937c-a9d91599c459.webm)

Pavneet-Sing commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -116,35 +117,68 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
-    std::move(callback).Run(std::move(blockchain_buy_tokens));
-    return;
+  std::vector<mojom::BlockchainToken> buy_tokens;
+  if (std::find(providers.begin(), providers.end(),
+                mojom::OnRampProvider::kWyre) != providers.end()) {
+    std::vector<mojom::BlockchainToken> wyre_buy_tokens = GetWyreBuyTokens();
+    buy_tokens.insert(buy_tokens.begin(), wyre_buy_tokens.begin(),
+                      wyre_buy_tokens.end());
```
wyre has been removed https://github.com/brave/brave-browser/issues/27678 so don't think we need this.

Pavneet-Sing commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -116,35 +117,68 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
   }
-
-  for (const auto& token : *buy_tokens) {
+  if (std::find(providers.begin(), providers.end(),
+                mojom::OnRampProvider::kRamp) != providers.end()) {
+    std::vector<mojom::BlockchainToken> ramp_buy_tokens = GetRampBuyTokens();
+    buy_tokens.insert(buy_tokens.begin(), ramp_buy_tokens.begin(),
+                      ramp_buy_tokens.end());
+  }
+  if (std::find(providers.begin(), providers.end(),
+                mojom::OnRampProvider::kSardine) != providers.end()) {
+    std::vector<mojom::BlockchainToken> sardine_buy_tokens =
+        GetSardineBuyTokens();
+    buy_tokens.insert(buy_tokens.begin(), sardine_buy_tokens.begin(),
+                      sardine_buy_tokens.end());
+  }
+  if (std::find(providers.begin(), providers.end(),
+                mojom::OnRampProvider::kTransak) != providers.end()) {
+    std::vector<mojom::BlockchainToken> transak_buy_tokens =
+        GetTransakBuyTokens();
+    buy_tokens.insert(buy_tokens.begin(), transak_buy_tokens.begin(),
+                      transak_buy_tokens.end());
+  }
```
nit: can be simplified with a loop.

Pavneet-Sing commented on components/brave_wallet/common/brave_wallet.mojom:
```diff
@@ -668,6 +668,7 @@ interface BlockchainRegistry {
   // Below APIs are Ethereum only for the moment.
   // Obtains all tokens for the Buy UI
   GetBuyTokens(OnRampProvider provider, string chain_id) => (array<BlockchainToken> tokens);
+  GetProvidersBuyTokens(array<OnRampProvider> providers, string chain_id) => (array<BlockchainToken> tokens);
```
nit: add comment docs

Pavneet-Sing reviewed (approved):
LG, [address the search list issue](https://github.com/brave/brave-core/pull/16872#issuecomment-1420399177) and comments for possible optimisations. Feel free to resolve once done.

Pavneet-Sing commented on browser/ui/android/strings/android_brave_strings.grd:
```diff
@@ -2035,7 +2035,7 @@ Are you sure you want to do this?
     </message>
     <message name="IDS_WALLET_BUY_MAINNET_BUTTON_TEXT" desc="Buy button on main net text">
-      Continue to Ramp
+      Select purchase method
```
Title case

simoarpe commented:
Addressed review feedback for Java parts @Pavneet-Sing.
All valid points for Core too, but before delving more into C++ code I'll wait for @yrliou's review as maybe there are further adjustments to make.

simoarpe commented on components/brave_wallet/browser/brave_wallet_constants.cc:
```diff
@@ -4,6 +4,7 @@
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include <map>
+#include <set>
```
Sorry, this is an unused code refuse from a previous commit where I tried a solution using sets. Solution that I reverted because it was too problematic.

yrliou replied:
^ cc @simoarpe this hasn't been removed yet

simoarpe commented on components/brave_wallet/browser/brave_wallet_constants.h:
```diff
@@ -6,6 +6,7 @@
 #ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_BRAVE_WALLET_CONSTANTS_H_
 #define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_BRAVE_WALLET_CONSTANTS_H_
 
+#include <set>
```
https://github.com/brave/brave-core/pull/16872#discussion_r1100018591

simoarpe commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -5,6 +5,7 @@
 
 #include "brave/components/brave_wallet/browser/blockchain_registry.h"
 
+#include <set>
```
https://github.com/brave/brave-core/pull/16872#discussion_r1100018591

simoarpe commented on components/brave_wallet/browser/blockchain_registry.h:
```diff
@@ -51,9 +51,16 @@ class BlockchainRegistry : public mojom::BlockchainRegistry {
   void GetAllTokens(const std::string& chain_id,
                     mojom::CoinType coin,
                     GetAllTokensCallback callback) override;
+  std::vector<brave_wallet::mojom::BlockchainTokenPtr> GetBuyTokens(
```
Sure!

simoarpe commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -116,35 +117,68 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
-void BlockchainRegistry::GetBuyTokens(mojom::OnRampProvider provider,
-                                      const std::string& chain_id,
-                                      GetBuyTokensCallback callback) {
+std::vector<brave_wallet::mojom::BlockchainTokenPtr>
```
And also for `brave_wallet::mojom::OnRampCurrencyPtr` I presume.

simoarpe commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -116,35 +117,68 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
-                                      const std::string& chain_id,
-                                      GetBuyTokensCallback callback) {
+std::vector<brave_wallet::mojom::BlockchainTokenPtr>
+BlockchainRegistry::GetBuyTokens(
```
Thanks @yrliou, handling it this way using a `flat_set` will help a lot. What caused me trouble were duplications related to contract addresses declared all uppercase for some providers, and all lowercase for others.
For example, check the Basic Attention Token on mainnet.
For the OnRamp.Network the contract address is declared all lowercase: https://github.com/brave/brave-core/blob/c73d16cc1f77ec351edd402552549f1d93980109/components/brave_wallet/browser/brave_wallet_constants.cc#L158-L160

While for Sardine it's all uppercase:
https://github.com/brave/brave-core/blob/c73d16cc1f77ec351edd402552549f1d93980109/components/brave_wallet/browser/brave_wallet_constants.cc#L359-L361

This of course led to duplication in the UI, and I've eventually implemented a method in the Java layer to remove duplicated entries. Not sure if there's an easy way to handle this at C++ level, if you have suggestions I'll be happy to include them. Otherwise, we can leave it as is and keep the duplication check in Java.

yrliou replied:
I don't think the duplicate tokens needs to be blocking this PR as other frontends are handling duplicates already too, we can open a follow-up issue to return a map (since mojo doesn't support set afaik) instead of array to help mitigate this all in backend.

simoarpe replied:
Fine by me 👍

yrliou commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -100,51 +101,64 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
+    const std::vector<mojom::OnRampProvider>& providers,
+    const std::string& chain_id,
+    GetProvidersBuyTokensCallback callback) {
+  std::vector<mojom::BlockchainTokenPtr> blockchain_buy_tokens =
+      GetBuyTokens(providers, chain_id);
   std::move(callback).Run(std::move(blockchain_buy_tokens));
```
Simplify to one line as above too.

yrliou commented on components/brave_wallet/browser/blockchain_registry_unittest.cc:
```diff
@@ -456,6 +456,27 @@ TEST(BlockchainRegistryUnitTest, GetBuyTokens) {
   run_loop4.Run();
 }
 
+TEST(BlockchainRegistryUnitTest, GetProvidersBuyTokens) {
```
This test is not enough, what you're doing here is only checking the first two mainnet tokens in Wyre buy tokens and didn't really cover the expected result, we should probably construct a list with all tokens we expected in the API result by calling Get*BuyTokens functions.

yrliou replied:
Also I would suggest a test case that have duplicate providers passed and make sure the returned token list is still the same as providers without duplicate.

yrliou commented on components/brave_wallet/browser/blockchain_registry.cc:
```diff
@@ -100,51 +101,64 @@ void BlockchainRegistry::GetAllTokens(const std::string& chain_id,
+      buy_tokens = &GetSardineBuyTokens();
+    } else if (provider == mojom::OnRampProvider::kTransak) {
+      buy_tokens = &GetTransakBuyTokens();
+    }
```
Please also add below to skip when buy_tokens is nullptr, thanks.
```
} else {
  continue;
}

simoarpe commented:
Ready for another review round @yrliou : [feedback addressed](https://github.com/brave/brave-core/pull/16872/commits/0a5c26a5926d10fd658aacdd8fcbca1b34a2c193), and [test improved](https://github.com/brave/brave-core/pull/16872/commits/a89e4c9340ca3ed6a7418cd74ff562b9e68f3fe9).

yrliou commented on components/brave_wallet/browser/blockchain_registry_unittest.cc:
```diff
@@ -460,21 +462,211 @@ TEST(BlockchainRegistryUnitTest, GetProvidersBuyTokens) {
 
-  // Get Wyre, Ramp, Sardine and Transak buy tokens.
-  base::RunLoop run_loop;
+  std::vector<mojom::BlockchainToken> buy_tokens;
```
```
TEST(BlockchainRegistryUnitTest, GetProvidersBuyTokens) {
  base::test::TaskEnvironment task_environment;
  auto* registry = BlockchainRegistry::GetInstance();

  std::vector<mojom::BlockchainToken> buy_tokens;
  for (const auto& v : {GetWyreBuyTokens(), GetRampBuyTokens(),
                        GetSardineBuyTokens(), GetTransakBuyTokens()}) {
    buy_tokens.insert(buy_tokens.end(), v.begin(), v.end());
  }

  const char* chains[] = {
      mojom::kMainnetChainId,          mojom::kPolygonMainnetChainId,
      mojom::kAvalancheMainnetChainId, mojom::kBinanceSmartChainMainnetChainId,
      mojom::kSolanaMainnet,           mojom::kCeloMainnetChainId,
      mojom::kArbitrumMainnetChainId};

  for (auto* chain : chains) {
    std::vector<mojom::BlockchainTokenPtr> expected_tokens;
    for (const auto& token : buy_tokens) {
      if (token.chain_id == chain) {
        expected_tokens.push_back(mojom::BlockchainToken::New(token));
      }
    }

    base::RunLoop run_loop;
    registry->GetProvidersBuyTokens(
        {mojom::OnRampProvider::kWyre, mojom::OnRampProvider::kRamp,
         mojom::OnRampProvider::kSardine, mojom::OnRampProvider::kTransak,
         mojom::OnRampProvider::kTransak /* test duplicate provider */},
        chain,
        base::BindLambdaForTesting(
            [&](std::vector<mojom::BlockchainTokenPtr> token_list) {
              EXPECT_NE(token_list.size(), 0UL);
              EXPECT_NE(expected_tokens.size(), 0UL);
              EXPECT_TRUE(std::equal(expected_tokens.begin(),
                                     expected_tokens.end(), token_list.begin(),
                                     token_list.end()))
                  << chain;

              run_loop.Quit();
            }));
    run_loop.Run();
  }
}
```

yrliou replied:
@simoarpe we can just write
`EXPECT_EQ(expected_tokens, token_list) << chain;`
no need to check both sizes and std::equal.

yrliou reviewed (approved):
core++ please squash commits into one before merging.

Pavneet-Sing commented on android/java/org/chromium/chrome/browser/crypto_wallet/adapters/WalletCoinAdapter.java:
```diff
@@ -241,11 +241,11 @@ public void setWalletCoinAdapterType(AdapterType type) {
 
     public void setWalletListItemModelList(List<WalletListItemModel> walletListItemModelList) {
-        this.walletListItemModelList = walletListItemModelList;
+        this.walletListItemModelList = removeDuplicates(walletListItemModelList);
```
I missed to notice this before but this should be done where the data is fetched or processed then passed the filtered data to UI. Ok for now as a workaround, probably address this in the next PR.

Pavneet-Sing replied:
It is causing an issue to show tokens from all networks so logged an issue https://github.com/brave/brave-browser/issues/28506
